viewing_party/party.py

```diff
@@ -0,0 +1,202 @@
+    
+    friends_movies = set()
+    for friends_data in user_data["friends"]:
+        for watched in friends_data["watched"]:
+            for title, movie_titles in watched.items():
+                friends_movies.add((title, movie_titles))
+
+    friends_unique = friends_movies - user_movies
+
+    uni_friends_movies = []
+    for movie in friends_unique:
+        x, y = movie
+        uni_friends_movies.append({x : y})
+    ##Why though??##
+    if len(uni_friends_movies) == 0:
+        return []
+
+    return uni_friends_movies
+
+#wave 4
+
+#user has not watched
+def user_has_watched(movie, watched_movies):
+    if movie in watched_movies:
+        return True
+    else:
+        return False 
+    
+    #at least 1 friend has watched it
+def in_friends_watchlist(movie, watched_by_friends):
+    if movie in watched_by_friends:
+        return True
+    else:
+        return False
+    #the host of the movie is a service that is a service in subscription
+def in_users_subcriptions(host, subscriptions):
+    if host in subscriptions:
+        return True
+    else:
+        return False
+
+def get_friends_movies(user_data):
+    titles = set()
+    friends = user_data["friends"]
+    movies = []
+    for friend in friends:
+        for movie in friend["watched"]:
+            if movie["title"] not in titles:
+                movies.append(movie)
+                titles.add(movie["title"])
+    return movies
+
+def get_available_recs(user_data):
+    friends_movies = get_friends_movies(user_data)
+    movie_recs = []
+
+    if len(friends_movies) == 0:
+        return []
+
+    titles_of_movies_user_watched = set()
+    for movie in user_data["watched"]:
+        titles_of_movies_user_watched.add(movie["title"])
+
+    for movie in friends_movies:
+        if movie["title"] not in titles_of_movies_user_watched:
+            if in_users_subcriptions(movie["host"], user_data["subscriptions"]):
+                movie_recs.append(movie)
+    return movie_recs
+
+
+##wave 5
+def get_new_rec_by_genre(user_data):
+    friends_movies = get_friends_movies(user_data)
+    
+    if len(user_data["watched"]) == 0:
+        return []
+
+    if len(friends_movies) == 0:
+        return []
+    
+    titles_of_movies_user_watched = set()
+    for movie in user_data["watched"]:
+        titles_of_movies_user_watched.add(movie["title"])
+    
+    recommended_by_genre = []
+    for movie in friends_movies:
+        if movie["title"] not in titles_of_movies_user_watched:
+            if movie["genre"] == get_most_watched_genre(user_data):
+                recommended_by_genre.append(movie)
+    return recommended_by_genre
+
+def get_rec_from_favorites(user_data):
+    if len(user_data["watched"]) and len(user_data["favorites"])== 0:
+        return []
+
+    friends_movies = get_friends_movies(user_data)
+    favorite_movies = []
+    for movie in user_data["favorites"]:
+        if movie not in friends_movies:
```

Comment:
It would be worthwhile to extract the movie titles from the `friend_movies` (a list) into a set of titles like you did elsewhere. Using the set to check whether the current movie has been seen by the friends would improve the `in` lookup here from O(n) to O(1).

Try the approach you used in your other functions that used sets of titles!